### PR TITLE
New native types

### PR DESCRIPTION
**Pull request by aolfat**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (non-breaking change that modifies some code)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

## Summary by CodeRabbit

- **New Features**
  - Added debug logging for primary table registration.
  - Introduced structured native type interfaces and implementations for consistent type representation.
  - Enhanced dataset schema retrieval with detailed native type support across multiple SQL providers.
  - Added row and value protobuf conversion methods for improved data serialization.
  - Extended numeric conversion functions to support all standard integer sizes.
  - Added nullable type support for ClickHouse native types.
  - Introduced precision and scale handling for Snowflake NUMBER types.
  - Added comprehensive mock implementations for unit testing online and offline stores.

- **Bug Fixes**
  - Fixed column name handling in source data and dashboard APIs for consistency.
  - Adjusted test expectations to match actual column name formats.

- **Refactor**
  - Major refactor of value conversion logic across all SQL providers using structured native type abstractions.
  - Unified dataset interfaces and removed legacy adapter patterns in offline store implementations.
  - Centralized offline store retrieval logic in serving layer with special handling for SparkOffline stores.
  - Updated dataset handling to use interface types instead of concrete structs for flexibility.
  - Simplified primary table row count retrieval logic.

- **Tests**
  - Expanded and modernized test coverage for native type parsing, value conversion, edge cases, and schema handling.
  - Updated test data to use typed native type constants instead of raw strings.
  - Added new tests for nullable types, precision/scale handling, and timezone equivalence.

- **Chores**
  - Improved import organization and code formatting in tests and source files.
  - Added detailed error wrapping and logging for schema retrieval and type conversion operations.

### CODE REVIEW

File: serving/serving.go [hunk 1]
```diff
@@ -321,8 +321,36 @@ func (serv *FeatureServer) SourceData(req *pb.SourceDataRequest, stream pb.Featu
...
+		serv.Logger.Debugw("Got source data row", "Values", values)
+		sRow, err = values.ToProto()
+
+		if isSpark {
+			vals := make([]interface{}, len(values))
+			for i, v := range values {
+				vals[i] = v.Value
+			}
+			sRow, err = SerializedSourceRow(vals)
+		}
```
[the remark this answers is not in the document]
Comment (aolfat) [reply serving/serving.go:352]: because we haven't implemented this for spark yet, will leave a comment

File: provider/correctness_test.go [hunk 2]
```diff
@@ -43,22 +44,22 @@ func TestTransformations(t *testing.T) {
...
-			"SELECT location_id, AVG(wind_speed) as avg_daily_wind_speed, AVG(wind_duration) as avg_daily_wind_duration, AVG(fetch_value) as avg_daily_fetch, DATE(timestamp) as date FROM %s GROUP BY location_id, DATE(timestamp)",
-		},
+		//{
+		//	getConfiguredBigQueryTester(t),
+		//	"SELECT location_id, AVG(wind_speed) as avg_daily_wind_speed, AVG(wind_duration) as avg_daily_wind_duration, AVG(fetch_value) as avg_daily_fetch, TIMESTAMP(timestamp) as date FROM %s GROUP BY location_id, TIMESTAMP(timestamp)",
+		//},
+		//{
+		//	getConfiguredSnowflakeTester(t),
+		//	"SELECT location_id, AVG(wind_speed) as avg_daily_wind_speed, AVG(wind_duration) as avg_daily_wind_duration, AVG(fetch_value) as avg_daily_fetch, DATE(timestamp) as date FROM %s GROUP BY location_id, DATE(timestamp)",
+		//},
```
Comment (ff-kamal) [inline provider/correctness_test.go:54]: Uncomment these.

File: provider/correctness_test.go [hunk 3]
```diff
@@ -43,22 +44,22 @@ func TestTransformations(t *testing.T) {
...
 			"SELECT LOCATION_ID, AVG(WIND_SPEED) as AVG_DAILY_WIND_SPEED, AVG(WIND_DURATION) as AVG_DAILY_WIND_DURATION, AVG(FETCH_VALUE) as AVG_DAILY_FETCH, DATE(TIMESTAMP) as DATE FROM %s GROUP BY LOCATION_ID, DATE(TIMESTAMP)",
 		},
-		{
-			getConfiguredClickHouseTester(t),
-			"SELECT LOCATION_ID, AVG(WIND_SPEED) as AVG_DAILY_WIND_SPEED, AVG(WIND_DURATION) as AVG_DAILY_WIND_DURATION, AVG(FETCH_VALUE) as AVG_DAILY_FETCH, DATE(TIMESTAMP) as DATE FROM %s GROUP BY LOCATION_ID, DATE(TIMESTAMP)",
-		},
+		//{
+		//	getConfiguredClickHouseTester(t),
+		//	"SELECT LOCATION_ID, AVG(WIND_SPEED) as AVG_DAILY_WIND_SPEED, AVG(WIND_DURATION) as AVG_DAILY_WIND_DURATION, AVG(FETCH_VALUE) as AVG_DAILY_FETCH, DATE(TIMESTAMP) as DATE FROM %s GROUP BY LOCATION_ID, DATE(TIMESTAMP)",
+		//},
```
Comment (ff-kamal) [inline provider/correctness_test.go:62]: Uncomment these.

File: provider/offline.go [hunk 4]
```diff
@@ -530,9 +530,9 @@ type OfflineStoreCore interface {
 
 type OfflineStoreDataset interface {
 	// CreatePrimaryTable is not used outside of the context of tests
-	CreatePrimaryTable(id ResourceID, schema TableSchema) (dataset.Dataset, error)
+	CreatePrimaryTable(id ResourceID, schema TableSchema) (dataset.Dataset, error) // won't be used later
 	RegisterPrimaryFromSourceTable(id ResourceID, tableLocation pl.Location) (dataset.Dataset, error)
-	GetPrimaryTable(id ResourceID, source metadata.SourceVariant) (dataset.Dataset, error)
+	GetPrimaryTable(id ResourceID, source metadata.SourceVariant) (dataset.Dataset, error) // rename to GetDatset
```
Comment (ff-kamal) [inline provider/offline.go:535]: Is the comment still relevant?

File: provider/snowflake/native_types.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+	DATE          = fftypes.NativeTypeLiteral("DATE")
+	DATETIME      = fftypes.NativeTypeLiteral("DATETIME")
+	TIME          = fftypes.NativeTypeLiteral("TIME")
+	TIMESTAMP     = fftypes.NativeTypeLiteral("TIMESTAMP")
+	TIMESTAMP_LTZ = fftypes.NativeTypeLiteral("TIMESTAMP_LTZ")
+	TIMESTAMP_NTZ = fftypes.NativeTypeLiteral("TIMESTAMP_NTZ")
+	TIMESTAMP_TZ  = fftypes.NativeTypeLiteral("TIMESTAMP_TZ")
+
+	NUMBER = NewNumberType()
```
Comment (ff-kamal) [inline provider/snowflake/native_types.go:38]: What's the context for this?

File: provider/snowflake_types_test.go [hunk 6]
```diff
@@ -38,61 +39,61 @@ func NewSnowflakeTestData(t *testing.T) TestColumnData {
...
+			//	VerifyFunc: func(t *testing.T, actual interface{}) {
+			//		assert.InDelta(t, 3.14159, actual.(float32), 0.0001, "Float value mismatch")
+			//	},
+			//},
+			//{
+			//	Name:           "float4_col",
+			//	NativeType:     snowflake.FLOAT4,
+			//	ExpectedGoType: fftypes.Float32,
+			//	TestValue:      2.71828,
+			//	VerifyFunc: func(t *testing.T, actual interface{}) {
```
Comment (ff-kamal) [inline provider/snowflake_types_test.go:63]: Uncomment these?

File: provider/snowflake_types_test.go [hunk 7]
```diff
@@ -161,14 +162,23 @@ func NewSnowflakeTestData(t *testing.T) TestColumnData {
...
 			},
+			//{
+			//	Name:           "number_col",
+			//	NativeType:     snowflake.NUMBER,
+			//	ExpectedGoType: fftypes.Float64,
+			//	TestValue:      123.456,
+			//	VerifyFunc: func(t *testing.T, actual interface{}) {
+			//		assert.InDelta(t, 123.456, actual.(float64), 0.0001, "NUMBER value mismatch")
+			//	},
+			//},
```
Comment (ff-kamal) [inline provider/snowflake_types_test.go:181]: Any reason we don't have this test enabled?

**Review by ff-kamal — APPROVED** [review]
